### PR TITLE
feat: embed json certificate into pdf

### PR DESCRIPTION
**Pull request by christophbuehler**
# Description

- [x] #249
- [x] #245
- [x] #244
- [x] #240

**TL;DR**
```TS
generatePdf(
  certificate,
  {
-   inputType: 'json',
-   outputType: 'buffer',
+   attachCertificate: true,
+   a3Compliant: true,
  }
)
```

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Refactor (a code change that neither fixes a bug nor adds a feature)
- [x] Dependencies update
- [ ] This change requires a documentation update

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Review by christophbuehler — COMMENTED** [review]
Self review completed.

File: packages/generate-pdf/test/generate-pdf.spec.ts [hunk 1]
```diff
@@ -248,6 +252,42 @@ const runPDFGenerationTests = (testSuite: PDFGenerationTestProperties) => {
       expect(resultHash).toEqual(expectedHash);
     }, 10000);
 
+    it('should attach the certificate as JSON document if `attachCertificate` is true', async () => {
+      const translations = JSON.parse(readFileSync(translationsPath, 'utf8'));
+      (attachFileToPdf as jest.Mock).mockClear();
```
Comment (getlarge) [inline packages/generate-pdf/test/generate-pdf.spec.ts:257]: ideally you could run `(attachFileToPdf as jest.Mock).mockClear();` in `afterEach` hook

File: packages/generate-pdf/src/index.ts [hunk 2]
```diff
@@ -241,9 +257,15 @@ export async function generatePdf(
   options: GeneratePdfOptions = {},
 ): Promise<Buffer | PDFKit.PDFDocument> {
   const opts: GeneratePdfOptions = options ? { ...generatePdfOptions, ...options } : generatePdfOptions;
+
+  if (opts.attachCertificate && opts.outputType === 'stream') {
+    throw new Error('Cannot attach certificate to a stream output type. Please use buffer output type.');
+  }
```
Comment (getlarge) [inline packages/generate-pdf/src/index.ts:263]: Even though this might not be very performant it is still possible to apply the attachCertificateToPdf and return a readable stream out of the buffer; hint Readable.from(YourBuffer).

File: packages/generate-pdf/src/attach-file-to-pdf.ts [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+  fileContent: string,
+  fileName: string,
+  options: AttachmentOptions,
+): Promise<Buffer> => {
+  const pdfDoc = await PDFDocument.load(pdfBuffer);
+  const encodedFileContent = Buffer.from(fileContent).toString('base64');
+  await pdfDoc.attach(encodedFileContent, fileName, options);
+  const uint8Array = await pdfDoc.save();
+  return Buffer.from(uint8Array);
+};
```
Comment (getlarge) [inline packages/generate-pdf/src/attach-file-to-pdf.ts:22]: Is the purpose of the PR only to allow attachment to the PDF or also to make it PDF/A3 compliant?
Too bad stream can't be used in this library, because that's potentially a lot of memory allocated in this function.

File: packages/generate-pdf/example_coa.ts [hunk 4]
```diff
@@ -51,29 +52,19 @@ const generatorPath = '../generate-coa-pdf-template/dist/generateContent.cjs';
...
-    await new Promise((resolve, reject) => {
-      writeStream
-        .on('finish', () => {
-          resolve(true);
-        })
-        .on('error', (err) => {
-          reject(err);
-        });
-    });
+    await writeFile(outputFilePath, pdfDoc);
```
Comment (getlarge) [inline packages/generate-pdf/example_coa.ts:65]: Maybe now is a good time to start measuring how this example performs?
For example using [`perf_hooks`](https://nodejs.org/api/perf_hooks.html#perf_hooksperformance) to check durations and `process. memoryUsage()` to track memory consumption (or even better with the heap profiler).

File: package.json [hunk 5]
```diff
@@ -91,6 +91,7 @@
     "lodash.merge": "^4.6.2",
     "lodash.set": "^4.3.2",
     "node-cache": "^5.1.2",
+    "pdf-lib": "1.17.1",
```
Comment (getlarge) [inline package.json:94]: i supposed you did your benchmark already, but that is surprising no recently maintained library enables PDF manipulation.

File: packages/generate-pdf/src/attach-file-to-pdf.ts (hunk 3 above)
Comment (christophbuehler) [reply packages/generate-pdf/src/attach-file-to-pdf.ts:22]: The primary goal is to allow attachments, but having a PDF/A3 document would be better. I am looking into that today, potentially using some different lib.

File: package.json (hunk 5 above)
Comment (christophbuehler) [reply package.json:94]: Yes, the library may change with adding PDF/A3 compliancy though.

File: packages/generate-pdf/src/index.ts (hunk 2 above)
Comment (christophbuehler) [reply packages/generate-pdf/src/index.ts:263]: Yes, but I do not want to encourage that. Just use a buffer if you want this feature. The conversion should be done on the caller side if a stream is required.

Comment (christophbuehler) [reply packages/generate-pdf/src/index.ts]: We have decided to remove the 'stream' output type all together.

File: packages/generate-pdf/example_coa.ts (hunk 4 above)
Comment (christophbuehler) [reply packages/generate-pdf/example_coa.ts]: In another PR we can do that. This one is too large already.

**Review by christophbuehler — COMMENTED** [review]
Replaced Lato with NotoSans, PDF A/3 compliance, cleanup

File: packages/generate-pdf/tsconfig.json [hunk 6]
```diff
@@ -8,9 +8,6 @@
   "references": [
     {
       "path": "./tsconfig.lib.json"
-    },
-    {
-      "path": "./tsconfig.spec.json"
     }
```
Comment (getlarge) [inline packages/generate-pdf/tsconfig.json:11]: I am curious: Why did you have to remove the TSconfig ref?

File: packages/generate-pdf/src/pdf-a-3a/index.ts [hunk 7]
```diff
@@ -0,0 +1,42 @@
+import { PDFDocument, PDFName } from 'pdf-lib';
+
+import { addStructTree } from './add-struct-tree';
+import { setColorProfile } from './set-color-profile';
+import { addXMPMetadata } from './xmp-metadata';
+
+export const AUTHOR = 'S1SEVEN';
+export const PRODUCER = 'S1SEVEN Schema Tools';
+export const COLOR_PROFILE_FILE = `${__dirname}/../../../../fixtures/color-profiles/sRGB2014.icc`;
```
Comment (getlarge) [inline packages/generate-pdf/src/pdf-a-3a/index.ts:9]: When using the compiled code, this probably won't resolve to the correct path.
Could you add this file to the generate-pdf package and copy it in the asset during compilation?

File: packages/generate-pdf/src/index.ts [hunk 8]
```diff
@@ -24,68 +23,136 @@ import {
...
-  // eslint-disable-next-line @typescript-eslint/no-explicit-any
-  (_module as any)._compile(code, fileName);
-  _module.loaded = true;
-  return _module.exports;
+const defaultFonts: TFontDictionary = {
+  NotoSans: {
+    normal: `${__dirname}/../../../fixtures/fonts/NotoSans-Regular.ttf`,
+    bold: `${__dirname}/../../../fixtures/fonts/NotoSans-Bold.ttf`,
+    italics: `${__dirname}/../../../fixtures/fonts/NotoSans-Italic.ttf`,
+  },
```
Comment (getlarge) [inline packages/generate-pdf/src/index.ts:56]: Are you sure those fonts can be loaded when running the compiled code?

Comment (christophbuehler) [reply packages/generate-pdf/src/index.ts:56]: We load fonts from the fixtures [in other places](https://github.com/s1seven/schema-tools/blob/main/utils/create-pdf-certificate.ts), but I am not sure how you make sure the fixtures are part of the output.

File: packages/generate-pdf/test/generate-pdf.spec.ts [hunk 9]
```diff
@@ -152,18 +157,6 @@ const generatePaths = (
   };
 };
 
-const waitWritableStreamEnd = (writeStream: Writable, outputFilePath: string) => {
-  return new Promise((resolve, reject) => {
-    writeStream
-      .on('finish', () => {
-        expect(existsSync(outputFilePath)).toEqual(true);
-        unlinkSync(outputFilePath);
```
Comment (eamon0989) [inline packages/generate-pdf/test/generate-pdf.spec.ts:160]: Removing this means that the output PDF files are no longer deleted. They are output to the root directory.

File: packages/generate-pdf/test/generate-pdf.spec.ts [hunk 10]
```diff
@@ -188,29 +181,26 @@ const runPDFGenerationTests = (testSuite: PDFGenerationTestProperties) => {
...
             languageFontMap,
+            a3Compliant: true,
           };
-      //
+
       const pdfDoc = await generatePdf(validCertificate, generatePdfOptions);
-      const writeStream = createWriteStream(outputFilePath);
-      pdfDoc.pipe(writeStream);
-      pdfDoc.end();
-      await waitWritableStreamEnd(writeStream, outputFilePath);
```
Comment (eamon0989) [inline packages/generate-pdf/test/generate-pdf.spec.ts:213]: `waitWritableStreamEnd` contained an `expect` clause that checked the existence of the newly created file. Removing it means that check has also been removed.

**Review by eamon0989 — APPROVED** [review]
lgtm

File: utils/create-pdf-certificate.ts [hunk 11]
```diff
@@ -1,25 +1,26 @@
...
 import { fileExists, normalizePath } from './helpers';
 
 const fonts = {
-  Lato: {
-    normal: './node_modules/lato-font/fonts/lato-normal/lato-normal.woff',
-    bold: './node_modules/lato-font/fonts/lato-bold/lato-bold.woff',
-    italics: './node_modules/lato-font/fonts/lato-light-italic/lato-light-italic.woff',
-    light: './node_modules/lato-font/fonts/lato-light/lato-light.woff',
+  NotoSans: {
+    normal: `${__dirname}/../../fixtures/fonts/NotoSans-Regular.ttf`,
```
Comment (eamon0989) [inline utils/create-pdf-certificate.ts:14]: I tried running this locally and the fonts were not found, I think the correct path is `${__dirname}/../fixtures/fonts/NotoSans-Regular.ttf`

**Review by eamon0989 — CHANGES_REQUESTED** [review]

File: utils/create-pdf-certificate.ts (hunk 11 above)
Comment (christophbuehler) [reply utils/create-pdf-certificate.ts]: You are right, I updated the path.

**Review by eamon0989 — APPROVED** [review]
